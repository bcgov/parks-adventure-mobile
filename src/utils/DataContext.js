import React from 'react'
import PropTypes from 'prop-types'
import { getParks, getLocation, getActivities, getFacilities } from './api'
import { defaultDistanceFilter, filterParks } from './helpers'

export const DataContext = React.createContext()

export function DataProvider({ children }) {
  const [parks, setParks] = React.useState([])
  const [location, setLocation] = React.useState()
  const [distance, setDistance] = React.useState(defaultDistanceFilter)
  const [searchTerm, updateSearchTerm] = React.useState('')
  const [activities, updateActivities] = React.useState([])
  const [originalActivities, setOriginalActivities] = React.useState([])
  const [facilities, updateFacilities] = React.useState([])
  const [originalFacilities, setOriginalFacilities] = React.useState([])
  const [filteredList, updateFilteredList] = React.useState([])

  React.useEffect(() => {
    async function loadData() {
      const parks = await getParks()
      setParks(parks)
      updateFilteredList(parks)

      const location = await getLocation()
      setLocation(location)

      const activities = await getActivities()
      updateActivities(activities)
      setOriginalActivities(JSON.parse(JSON.stringify(activities)))

      const facilities = await getFacilities()
      updateFacilities(JSON.parse(JSON.stringify(facilities)))
      setOriginalFacilities(facilities)
    }
    loadData()
  }, [])

  React.useEffect(() => {
    applyFilters()
  }, [parks, location, distance, activities, facilities])

  function changeActivitySelectionState(id) {
    const index = activities.findIndex((activity) => activity.id === id)
    const newActivityList = [...activities]
    newActivityList[index].selected = !newActivityList[index].selected
    updateActivities(newActivityList)
  }

  function changeFacilitySelectionState(id) {
    const index = facilities.findIndex((facility) => facility.id === id)
    const newFacilityList = [...facilities]
    newFacilityList[index].selected = !newFacilityList[index].selected
    updateFacilities(newFacilityList)
  }

  function applyFilters(searchString) {
    if (searchString !== undefined) {
      updateSearchTerm(searchString)
    }

    const results = filterParks({
      parks,
      searchTerm: searchString === undefined ? searchTerm : searchString,
      location,
      distance,
      activities,
      facilities,
    })
    updateFilteredList(results)
  }

  function resetFilters() {
    setDistance(defaultDistanceFilter)
    updateActivities(originalActivities)
    updateFacilities(originalFacilities)
  }

  function favoritePark(id) {
    const index = parks.findIndex((park) => park.id === id)
    const newList = [...parks]
    newList[index].favorited = !parks[index].favorited
    setParks(newList)
  }

  return (
    <DataContext.Provider
      value={{
        parks,
        location,
        distance,
        setDistance,
        activities,
        updateActivities: changeActivitySelectionState,
        facilities,
        updateFacilities: changeFacilitySelectionState,
        applyFilters,
        resetFilters,
        filteredList,
        filterApplied: parks.length - filteredList.length > 0,
        favoritePark,
      }}>
      {children}
    </DataContext.Provider>
  )
}

DataProvider.propTypes = {
  children: PropTypes.array,
}
