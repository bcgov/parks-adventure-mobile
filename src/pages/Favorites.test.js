import React from 'react'
import { create, act } from 'react-test-renderer'
import FavoritesPage from './Favorites'
import { DataProvider } from '../utils/DataContext.mock'

test('Favorites page matches snapshot', () => {
  let tree

  act(() => {
    tree = create(
      <DataProvider>
        <FavoritesPage navigation={{}} />
      </DataProvider>
    )
  })

  expect(tree.toJSON()).toMatchSnapshot()
})
