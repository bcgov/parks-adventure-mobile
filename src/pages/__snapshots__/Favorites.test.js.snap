// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Favorites page matches snapshot 1`] = `
<View
  style={
    Array [
      Object {
        "backgroundColor": "white",
        "flexBasis": 0,
        "flexGrow": 1,
        "flexShrink": 1,
      },
    ]
  }
>
  <View
    locationNotAvailable={false}
    style={
      Array [
        Object {
          "flexBasis": 0,
          "flexGrow": 1,
          "flexShrink": 1,
          "marginBottom": 0,
          "marginLeft": 20,
          "marginRight": 20,
          "marginTop": 0,
        },
      ]
    }
  >
    <RCTScrollView
      className="sc-bdnylx hmhEvm"
      contentContainerStyle={
        Object {
          "paddingBottom": 10,
        }
      }
      data={
        Array [
          Object {
            "activities": Array [],
            "distance": "383",
            "facilities": Array [],
            "favorited": true,
            "id": "361",
            "location": Object {
              "latitude": "50.987578",
              "longitude": "-119.725971",
            },
            "searchableTitle": "Adams Lake Provincial Park - Bush Creek Site",
            "title": "Adams Lake Provincial Park - Bush Creek Site",
          },
        ]
      }
      disableVirtualization={false}
      getItem={[Function]}
      getItemCount={[Function]}
      horizontal={false}
      initialNumToRender={10}
      keyExtractor={[Function]}
      maxToRenderPerBatch={10}
      onContentSizeChange={[Function]}
      onEndReachedThreshold={2}
      onLayout={[Function]}
      onMomentumScrollEnd={[Function]}
      onScroll={[Function]}
      onScrollBeginDrag={[Function]}
      onScrollEndDrag={[Function]}
      removeClippedSubviews={false}
      renderItem={[Function]}
      scrollEventThrottle={50}
      showsVerticalScrollIndicator={false}
      stickyHeaderIndices={Array []}
      updateCellsBatchingPeriod={50}
      viewabilityConfigCallbackPairs={Array []}
      windowSize={21}
    >
      <View>
        <View
          onLayout={[Function]}
          style={null}
        >
          <View
            accessibilityLabel="navigate to park details"
            accessible={true}
            focusable={true}
            onBlur={[Function]}
            onClick={[Function]}
            onFocus={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
          >
            <View
              className="sc-gtssRu MnPZe"
              noTopBorder={false}
            >
              <Image
                className="sc-dlnjPT fmUCMd"
                defaultSource={1}
                source={1}
              />
              <View
                className="sc-jSFkmK glLnMI"
              >
                <View
                  className="sc-gKAblj bdPEsA"
                >
                  <Text
                    className="sc-hKFyIo jvWbbP"
                    style={
                      Array [
                        Object {
                          "color": "#000000",
                          "fontFamily": "System",
                          "fontWeight": "400",
                          "textAlign": "left",
                        },
                        undefined,
                      ]
                    }
                  >
                    Adams Lake Provincial Park - Bush Creek Site
                  </Text>
                  <View
                    accessible={true}
                    focusable={true}
                    hitSlop={
                      Object {
                        "bottom": 50,
                        "left": 50,
                        "right": 50,
                        "top": 50,
                      }
                    }
                    onBlur={[Function]}
                    onClick={[Function]}
                    onFocus={[Function]}
                    onResponderGrant={[Function]}
                    onResponderMove={[Function]}
                    onResponderRelease={[Function]}
                    onResponderTerminate={[Function]}
                    onResponderTerminationRequest={[Function]}
                    onStartShouldSetResponder={[Function]}
                    style={
                      Object {
                        "margin": 12,
                      }
                    }
                  >
                    <View
                      accessibilityLabel="favorite park"
                      name="heart"
                      size={24}
                    />
                  </View>
                </View>
                <Text
                  className="sc-eCApGN fFLSlQ"
                  style={
                    Array [
                      Object {
                        "color": "#000000",
                        "fontFamily": "System",
                        "fontWeight": "400",
                        "textAlign": "left",
                      },
                      undefined,
                    ]
                  }
                >
                  383KM AWAY
                </Text>
              </View>
            </View>
          </View>
        </View>
      </View>
    </RCTScrollView>
  </View>
</View>
`;
